File: fries-leetcode-training/01-array-hash-table/01-two-sum.cpp
/**
 * LeetCode 1. 两数之和
 * https://leetcode.cn/problems/two-sum/
 * 
 * 给定一个整数数组nums和一个整数目标值target，请你在该数组中找出和为目标值target的那两个整数。
 * 
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> hashMap;
        
        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];
            
            if (hashMap.count(complement)) {
                return {hashMap[complement], i};
            }
            
            hashMap[nums[i]] = i;
        }
        
        return {};
    }
};

int main() {
    Solution solution;
    
    vector<pair<vector<int>, int>> testCases = {
        {{2, 7, 11, 15}, 9},
        {{3, 2, 4}, 6},
        {{3, 3}, 6}
    };
    
    for (const auto& [nums, target] : testCases) {
        cout << "数组: ";
        for (int num : nums) cout << num << " ";
        cout << ", 目标: " << target << endl;
        
        vector<int> result = solution.twoSum(const_cast<vector<int>&>(nums), target);
        
        cout << "结果: [";
        for (size_t i = 0; i < result.size(); i++) {
            cout << result[i];
            if (i < result.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
        
        if (result.size() == 2) {
            cout << "验证: " << nums[result[0]] << " + " << nums[result[1]] 
                 << " = " << nums[result[0]] + nums[result[1]] << endl;
        }
        cout << endl;
    }
    
    return 0;
}

File: fries-leetcode-training/01-array-hash-table/01-two-sum.py
"""
LeetCode 1. 两数之和
https://leetcode.cn/problems/two-sum/

给定一个整数数组nums和一个整数目标值target，请你在该数组中找出和为目标值target的那两个整数，并返回它们的数组下标。

哈希表解法

时间复杂度：O(n)
空间复杂度：O(n)
"""

def two_sum(nums, target):
    """
    两数之和 - 哈希表解法
    """
    hash_map = {}
    
    for i, num in enumerate(nums):
        complement = target - num
        
        if complement in hash_map:
            return [hash_map[complement], i]
        
        hash_map[num] = i
    
    return []

def two_sum_brute_force(nums, target):
    """
    两数之和 - 暴力解法
    """
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    return []

# 测试
if __name__ == "__main__":
    test_cases = [
        ([2, 7, 11, 15], 9),
        ([3, 2, 4], 6),
        ([3, 3], 6)
    ]
    
    for nums, target in test_cases:
        result = two_sum(nums, target)
        print(f"数组: {nums}, 目标: {target}")
        print(f"结果: {result}")
        print(f"验证: {nums[result[0]]} + {nums[result[1]]} = {nums[result[0]] + nums[result[1]]}\n")

